monai/optimizers/lr_finder.py

```diff
@@ -0,0 +1,629 @@
+                scale; otherwise, plotted in a linear scale.
+            ax: the plot is created in the specified matplotlib axes object and the
+                figure is not be shown. If `None`, then the figure and axes object are
+                created in this method and the figure is shown.
+            steepest_lr: plot the learning rate which had the steepest gradient.
+
+        Returns:
+            The matplotlib.axes.Axes object that contains the plot
+        """
+        if not has_matplotlib:
+            raise RuntimeError("Matplotlib is missing, can't plot result")
+
+        lrs, losses = self.get_lrs_and_losses(skip_start, skip_end)
+
+        # Create the figure and axes object if axes was not already given
+        fig = None
+        if ax is None:
+            fig, ax = plt.subplots()
+
+        # Plot loss as a function of the learning rate
+        ax.plot(lrs, losses)
+
+        # Plot the LR with steepest gradient
+        if steepest_lr:
+            lr_at_steepest_grad, loss_at_steepest_grad = self.get_steepest_gradient(skip_start, skip_end)
+            if lr_at_steepest_grad is not None:
+                ax.scatter(
+                    lr_at_steepest_grad,
+                    loss_at_steepest_grad,
+                    s=75,
+                    marker="o",
+                    color="red",
+                    zorder=3,
+                    label="steepest gradient",
+                )
+                ax.legend()
+
+        if log_lr:
+            ax.set_xscale("log")
+        ax.set_xlabel("Learning rate")
+        ax.set_ylabel("Loss")
+
+        # Show only if the figure was created internally
+        if fig is not None:
+            plt.show()
+
+        return ax
+
+
+class _LRSchedulerMONAI(_LRScheduler):
+    def __init__(self, optimizer: Optimizer, end_lr: float, num_iter: int, last_epoch: int = -1) -> None:
+        """
+        Args:
+            optimizer: wrapped optimizer.
+            end_lr: the final learning rate.
+            num_iter: the number of iterations over which the test occurs.
+            last_epoch: the index of last epoch.
+        Returns:
+            None
+        """
+        self.end_lr = end_lr
+        self.num_iter = num_iter
+        super(_LRSchedulerMONAI, self).__init__(optimizer, last_epoch)
+
+
+class LinearLR(_LRSchedulerMONAI):
+    """Linearly increases the learning rate between two boundaries over a number of
+    iterations.
+    """
+
+    def get_lr(self):
+        r = self.last_epoch / (self.num_iter - 1)
+        return [base_lr + r * (self.end_lr - base_lr) for base_lr in self.base_lrs]
+
+
+class ExponentialLR(_LRSchedulerMONAI):
+    """Exponentially increases the learning rate between two boundaries over a number of
+    iterations.
+    """
+
+    def get_lr(self):
+        r = self.last_epoch / (self.num_iter - 1)
+        return [base_lr * (self.end_lr / base_lr) ** r for base_lr in self.base_lrs]
+
+
+class StateCacher(object):
```

Comment:
I feel this is useful as a general utility, perhaps move it to monai.utils and add some docstrings? perhaps `state_dict` could be `data_obj`? this class doesn't assume a specific `state`